Improved DiffList

antis81 opened the pull request:
The DiffList now uses a ChangeListEntry struct instead of single variables in the callback. not sure about the `QVector<ChangeListEntry> DiffList::changes()` method. I'll try that out in one of the next commits.

Going further: Thinking about the PatchConsumer, it uses the same variables in the startFile() callback. It could use the same struct here, right? If so, I would like to change it in this PR.

Btw.: There was a little bug in the changeListCallBack() method a wrong handled bool. That caused only the first change to be noticed and returned an error after that -> Fixed it.

antis81 committed:
added ChangeListEntry struct

antis81 committed:
changed virtual method raw() to use the ChangeListEntry

scunz commented:
I agree, that `PatchConsumer::startFile` is the same callback as issued from `ChangeListConsumer`.
But thinking further about this, isn't it actually like this: A `PatchConsumer` is a more detailed version of a `ChangeListConsumer`, which would translate to C++ as `PatchConsumer` should inherit `ChangeListConsumer`?

antis81 committed:
added missing exports

antis81 committed:
added internal implementation of a DiffList::ChangeListConsumer

antis81 commented:
Yeah, inheritance seems to be really a good option. I added the rest of the changes to be able to call DiffList::changes(). Are you ok with that?

scunz commented on libGitWrap/DiffList.hpp:
```diff
@@ -24,13 +24,17 @@
     namespace Internal
     {
         class DiffListPrivate;
+        class DiffListConsumer;
```
actually not needed as forward decl, is it?

antis81 replied:
Yes, this one is obsolete. I'll remove it.

scunz commented on libGitWrap/DiffList.hpp:
```diff
@@ -24,13 +24,17 @@
+        class DiffListConsumer;
     }
 
+    typedef QVector<ChangeListEntry> ChangesList;
```
Shouldn't it be singular? i.e. `ChangeList`?

antis81 replied:
Probably would be better in singular.

scunz commented on libGitWrap/DiffList.cpp:
```diff
@@ -264,4 +293,12 @@
         return result;
     }
 
+    ChangesList DiffList::changes(Result &result)
```
Thinking again about my own naming: would be more consistent if we named it `changeList()` (like the result code)

antis81 replied:
Ok :smile:.

scunz commented:
Yepp, the general change is nice this way. (Just a few nitpicks amended).

antis81 committed:
multiple renames and cleanup

antis81 committed:
changed PatchConsumer being inherited from ChangeListConsumer; note that GIT_ERROR and GIT_OK also changed positions to match ChangeListConsumer

antis81 commented:
Another review, please. I implemented the inherited version of the PatchConsumer (it changes libMacGitverCore as well).

antis81 committed:
renamed methods "raw" and "startHunk" to improve code readability

antis81 commented:
:question: @scunz Unfortunately, there's documentation missing on lg2 about the `git_index_entry` struct. Maybe you have a better idea what the members mean:
- git_index_time ctime
- git_index_time mtime
  Is it creation and modified date?
- unsigned int dev
- unsigned int ino
- unsigned int mode
- unsigned int uid
- unsigned int gid
- unsigned short flags
- unsigned short flags_extended

Clear to me are these:
- char *path
- git_off_t file_size
- git_oid oid

The current (test)implementation simply runs through the index with a for loop and outputs the current entry's path. It actually shows all files in HEAD, wether they are staged or not.

# Edit: Moved this discussion to PR #10.

scunz commented on libGitWrap/IndexEntry.cpp:
```diff
@@ -0,0 +1,125 @@
+QTime IndexEntry::mTime() const
+{
+    git_index_time t = d->mEntry->mtime;
+    return QTime(0, 0, t.seconds );
```
Just a gut feeling without deeper research: isn't [QDateTime::setTime_t](http://doc.qt.digia.com/stable/qdatetime.html#setTime_t) the correct function to convert these?

antis81 replied:
Well, the main problem here is, that git_time is a uint64, but QDateTime and time_t use uint fields. This might be wrong for some 32-bit systems :bomb:.

scunz replied:
Hm, but the resolution is _seconds_. Meaning it will wrap around the 32bit border 136 _YEARS_ after 1.1.1970. Am I miscalculating?

scunz commented:
@antis81, could you separate this into 2 topics: One for the ChangeList changes and one for the Index.

I think, the ChangeList-Thingie looks more or less finished and could be merged; but related to the IndexEntry additions, I see a big issue with object life times, which I'd rather like to address without the ChangeList changes being intermingled into the diff views... :-)

antis81 commented:
Already thought about splitting branches. So, if you don't mind I would move the `topic/ImprovedDiffList` branch (edit: two or three commits before) and set up a new one based on current `development` (edit: for the index).

scunz commented:
Yes, that sounds like a plan :-)

antis81 commented:
Done.

scunz commented:
Cool, so you might now merge this one here (and it's associates) and be done with it :-)
